cmd/object-api-common.go

```diff
@@ -144,6 +145,75 @@ func cleanupDir(ctx context.Context, storage StorageAPI, volume, dirPath string)
 	return err
 }
 
+// Cleanup a directory recursively.
+func cleanupDirBulk(ctx context.Context, storage StorageAPI, volume string, dirPaths []string, errs []error) ([]error, error) {
+
+	var objsToDelete []string
+	var objsToDeleteIndexes []int
+
+	var traverse func(string) ([]string, error)
+	traverse = func(entryPath string) ([]string, error) {
+		var output = make([]string, 0)
+		if !hasSuffix(entryPath, slashSeparator) {
+			output = append(output, entryPath)
+			return output, nil
+		}
+		entries, err := storage.ListDir(volume, entryPath, -1, "")
+		if err != nil {
+			if err == errFileNotFound {
+				return nil, nil
+			}
+			return nil, err
+		}
+
+		for _, entry := range entries {
+			subEntries, err := traverse(pathJoin(entryPath, entry))
+			if err != nil {
+				return nil, err
+			}
+			output = append(output, subEntries...)
+		}
+		return output, nil
+	}
+
+	for idx, dirPath := range dirPaths {
+		if errs[idx] != nil {
+			continue
+		}
+		output, err := traverse(dirPath)
+		if err != nil {
+			errs[idx] = err
+			continue
+		} else {
+			errs[idx] = nil
+		}
+		objsToDelete = append(objsToDelete, output...)
+		for i := 0; i < len(output); i++ {
+			objsToDeleteIndexes = append(objsToDeleteIndexes, idx)
+		}
+	}
+
+	// Reverse the list so remove can succeed
+	sort.SliceStable(objsToDelete, func(i, j int) bool {
```

Comment:
sort.Slice is good enough, sort.SliceStable is needed for a completely different scenario. Also you should be using sort.Reverse()

Reply:
> sort.Slice is good enough, sort.SliceStable is needed for a completely different scenario. Also you should be using sort.Reverse()

I added a function which does the reverse without using sort, since it is confusing